### PR TITLE
Rustdoc: suggest removing disambiguator if linking to field

### LINKED ISSUE #85615: rustdoc ICE: "tried to calculate a disambiguator for a def without a namespace?"
Unfortunately I don't have an easy-to-repro case. This happened while working on Rocket's docs. I renamed a method which made a previously valid link invalid and I got this ICE. Fixing the doc link fixes the error. Backtrace from rustc 1.54.0-nightly (5dc8789e3 2021-05-21) follows:

```
warning: unresolved link to `Context::value`
  --> core/lib/src/form/context.rs:63:10
   |
63 | /// via [`Context::value()`] or [`Context::values()`]. Data fields do not have
   |          ^^^^^^^^^^^^^^^^^^ the struct `Context` has no function named `value`
   |
   = note: `#[warn(rustdoc::broken_intra_doc_links)]` on by default

thread 'rustc' panicked at 'tried to calculate a disambiguator for a def without a namespace?', src/librustdoc/passes/collect_intra_doc_links.rs:1607:18
stack backtrace:
   0: rust_begin_unwind
             at /rustc/5dc8789e300930751a78996da0fa906be5a344a2/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /rustc/5dc8789e300930751a78996da0fa906be5a344a2/library/core/src/panicking.rs:92:14
   2: core::option::expect_failed
             at /rustc/5dc8789e300930751a78996da0fa906be5a344a2/library/core/src/option.rs:1241:5
   3: rustdoc::passes::collect_intra_doc_links::suggest_disambiguator
   4: core::ops::function::FnOnce::call_once{{vtable.shim}}
   5: rustc_middle::lint::struct_lint_level::struct_lint_level_impl
   6: rustdoc::passes::collect_intra_doc_links::LinkCollector::resolve_link::{{closure}}
   7: rustdoc::passes::collect_intra_doc_links::LinkCollector::resolve_link::{{closure}}
   8: rustdoc::passes::collect_intra_doc_links::LinkCollector::resolve_link
   9: <rustdoc::passes::collect_intra_doc_links::LinkCollector as rustdoc::fold::DocFolder>::fold_item
  10: alloc::vec::source_iter_marker::<impl alloc::vec::spec_from_iter::SpecFromIter<T,I> for alloc::vec::Vec<T>>::from_iter
  11: rustdoc::fold::DocFolder::fold_inner_recur
  12: rustdoc::fold::DocFolder::fold_item_recur
  13: <rustdoc::passes::collect_intra_doc_links::LinkCollector as rustdoc::fold::DocFolder>::fold_item
  14: alloc::vec::source_iter_marker::<impl alloc::vec::spec_from_iter::SpecFromIter<T,I> for alloc::vec::Vec<T>>::from_iter
  15: rustdoc::fold::DocFolder::fold_inner_recur
  16: rustdoc::fold::DocFolder::fold_item_recur
  17: <rustdoc::passes::collect_intra_doc_links::LinkCollector as rustdoc::fold::DocFolder>::fold_item
  18: alloc::vec::source_iter_marker::<impl alloc::vec::spec_from_iter::SpecFromIter<T,I> for alloc::vec::Vec<T>>::from_iter
  19: rustdoc::fold::DocFolder::fold_inner_recur
  20: rustdoc::fold::DocFolder::fold_item_recur
  21: <rustdoc::passes::collect_intra_doc_links::LinkCollector as rustdoc::fold::DocFolder>::fold_item
  22: rustdoc::passes::collect_intra_doc_links::collect_intra_doc_links
  23: rustdoc::core::run_global_ctxt
  24: rustc_interface::passes::QueryContext::enter
  25: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  26: rustc_span::with_source_map
  27: rustc_interface::interface::create_compiler_and_run
  28: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

### PR DESCRIPTION
**Pull request by fee1-dead**
This fixes #85615.

@rustbot label T-rustdoc

### CODE REVIEW

File: src/test/rustdoc-ui/intra-doc/field-ice.stderr [hunk 1]
```diff
@@ -0,0 +1,15 @@
+error: incompatible link kind for `Foo::bar`
+  --> $DIR/field-ice.rs:5:6
+   |
+LL | /// [`Foo::bar()`] 
+   |      ^^^^^^^^^^^^ help: to link to the field, remove the disambiguator: ``Foo::bar``
```
Comment (GuillaumeGomez) [inline src/test/rustdoc-ui/intra-doc/field-ice.stderr:5]: The message is really not obvious. I think we should create a new error type so you can generate an error message looking like this:

```
help: to link to the field, remove the parenthesis: `Foo::bar`
```

(without the extra backquotes too)

Comment (fee1-dead) [reply src/test/rustdoc-ui/intra-doc/field-ice.stderr:5]: the prefix and suffix are all mapped to Disambiguator::Kind(DefKind::Fn):

`fn@foo`, `function@foo`, `method@foo`, and `foo()`.

There are two ways to approach this, either we only allow one kind of those four to exist, or we need to refactor to include that information inside `Disambiguator`.

**Comment by jyn514** [comment]
r=me with CI passing - I agree the diagnostic could be a little better, but that's a pre-existing bug, and this is already way better than the ICE.

@bors delegate=fee1-dead (please use r=jyn514, not r+)

File: src/test/rustdoc-ui/intra-doc/field-ice.stderr (hunk 1 above)
Comment (jyn514) [reply src/test/rustdoc-ui/intra-doc/field-ice.stderr:5]: > either we only allow one kind of those four to exist

This is a breaking change, it's not allowed

**Pull request merged**